cozy/synthesis/core.py

```diff
@@ -480,6 +480,15 @@ def _consider_replacement(
     print(" * in {}".format(pprint(target), pprint(e), pprint(replacement)))
     print(" * replacing {}".format(pprint(e)))
     print(" * with {}".format(pprint(replacement)))
+    from cozy.structures.treemultiset import ETreeMultisetElems
+    if isinstance(e, ETreeMultisetElems) and isinstance(e.e, EStateVar) and \
+            isinstance(replacement, EStateVar) and isinstance(replacement.e, ETreeMultisetElems):
+        # FIXME(zhen): current enumerator will always try to make ETreeMultisetElems a state var
+        # FIXME(zhen): we don't want this because we need to put TreeSet into state var, rather than its iterator
+        # FIXME(zhen): I still don't know how to fix this in a sensible way, but giving up an "improvement"
+        # FIXME(zhen): should be okay temporarily
+        print("give up {} -> {}".format(pprint(e), pprint(replacement)))
```

Comment:
This is a deficiency of Cozy's cost model, I think. The cost model incorrectly believes that storing the result of an `ETreeMultisetElems` on the data structure will be very cheap. In reality, maintaining that list is much much harder than maintaining the tree set itself.

The fix you have here is fine, but we should find a better solution at some point.